src/uucore/src/lib/features/entries.rs

```diff
@@ -73,14 +76,14 @@ pub fn get_groups() -> IOResult<Vec<gid_t>> {
     if ngroups == -1 {
         return Err(IOError::last_os_error());
     }
-    let mut groups = Vec::with_capacity(ngroups as usize);
+    let mut groups = vec![0; ngroups.try_into().unwrap()];
     let ngroups = unsafe { getgroups(ngroups, groups.as_mut_ptr()) };
     if ngroups == -1 {
         Err(IOError::last_os_error())
     } else {
-        unsafe {
-            groups.set_len(ngroups as usize);
-        }
+        let ngroups = ngroups.try_into().unwrap();
+        assert!(ngroups <= groups.len());
+        groups.truncate(ngroups);
```

Comment:
Can you please elaborate why you want to change this?

Reply:
To have less unsafe code. I think it's good practice not to use unsafe unless necessary, and in this case it didn't add much.

This way it's easy to verify that there's no uninitialized memory exposed and no buffer over-read possible.

I suspect the old version had a potential buffer over-read if the number of groups increased between calls to `getgroups()`. With the new code that just causes the assertion to fail, I don't think it's worth handling properly.

Reply:
> I suspect the old version had a potential buffer over-read if the number of groups increased between calls to `getgroups()`. With the new code that just causes the assertion to fail, I don't think it's worth handling properly.



According to the [docs](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.set_len) it’s safe to call `set_len()` if the elements were initialised and `new_len` <= `capacity`. (The docs also mention this exact use-case: "This method can be useful for situations in which the vector is serving as a buffer for other code, particularly over FFI".)

The old version complies to these safety restrictions because according to the manual: 
`int getgroups(int gidsetsize, gid_t grouplist[]);`
"If the number of groups of which a process is a member exceeds gidsetsize, getgroups() returns an error (EINVAL)."

I don't think the above lines are a safety issues that needs fixing.

FYI:
The [nix-rust](https://github.com/nix-rust/nix/blob/f9d508fe319ab394071cf70f9167517b5217b043/src/unistd.rs) crate also uses `with_capacity()` and `set_len()` in it’s FFI implementation for calls to `get groups()`, etc.

Reply:
What if `gidsetsize` is 0 because the user had no supplementary groups when the first call was made? My man page suggests it wouldn't look at the pointer.

(This does mean the assertion is useless, so I'll remove it. Thanks.)

Either way, my main reason for this change was simply to not have unsafe code that offers no benefits over safe code. That might make the PR title too strongly worded, but I think it's worthwhile.

Reply:
> What if `gidsetsize` is 0 because the user had no supplementary groups when the first call was made? 

That's a good point, there needs to be better return value handling, i.e.
* return early with an empty Vec<> if `getgroups` returns 0, so the "special" call to `getgroups` with gidsetsize=0 only happens once
* handle an EINVAL error by increasing the buffer and calling `getgroups` again...

Reply:
@jhscheer is that ok with you now ? :)
thanks

Reply:
1. The return value handling looks good, except that in case of an error, `vec![]` is called multiple times. I haven't looked into if using `resize()` here like in src/uucore/src/lib/features/entries.rs#L219 results in less allocations.
2. I'm not happy with replacing `with_capacity()` and `set_len()` in the name of "safety issues" without discussing potential efficiency trade-offs.
It may be that there's no difference because of compiler optimizations, however I doubt that since `with_capacity()` and `set_len()` are recommended in the docs for this use case.
In my opinion these operations are neither "unsafe" (because they operate within the safety restrictions) nor "unnecessary" (because they are more efficient).

Reply:
> except that in case of an error, `vec![]` is called multiple times.

It only happens in case of a very unlikely race condition so I simply didn't think it was worth the trouble.

I can change it if you think it's worth it for the sake of good practice but I don't think there's a meaningful efficiency gain.

> I haven't looked into if using `resize()` here like in src/uucore/src/lib/features/entries.rs#L219 results in less allocations.

I think it'd result in a call to `realloc` which can be cheaper than a full new allocation.

The original use of `resize()` in that location is actually a bit strange because that does zero-initialize it while `with_capacity()` didn't. I guess it was used because `reserve()` is trickier to work with.

> I'm not happy with replacing `with_capacity()` and `set_len()` in the name of "safety issues" without discussing potential efficiency trade-offs.

Sure, I've thought about those.

In a quick and dirty benchmark (on Debian) the difference between allocating a buffer vs allocating + initializing it seems to average around 30 nanoseconds (for 15 groups, which [historically](https://www.j3e.de/ngroups.html) pushed into the maximum on some systems though my FreeBSD VM now says 1023). The total cost of the syscalls to first get the number of groups and then fill the buffer varies from 1000 nanoseconds in the best case to 2500 nanoseconds in the worst case.

The function is called (indirectly) by `groups` and `id`. `groups` calls it at most once in its runtime, while `id` calls it either once or once for each of its arguments.

If you called `id` on a hundred users with over a hundred groups each it'd maybe cost 30 microseconds extra in total. And `id` takes a millisecond just to show its version information.

> It may be that there's no difference because of compiler optimizations

I also doubt that, rustc can't look inside the [kernel](https://github.com/torvalds/linux/blob/6880fa6c56601bb8ed59df6c30fd390cc5f6dd8f/kernel/groups.c#L44). I fully believed from the start that this had a cost, just that the cost was negligible.

> In my opinion these operations are neither "unsafe" (because they operate within the safety restrictions) nor "unnecessary" (because they are more efficient).

The thing is, it *was* unsafe. It had a memory safety issue that would have been prevented by only using safe `Vec` operations. I sure hope I'd get it right this time, but who knows?

I'll gladly do the `set_len` thing for a large buffer, but there's a tradeoff, and I don't see it being worth it in this case. It's only very marginally more efficient.

(I'm actually more worried about `uid2usr` and `usr2uid` and `grp2gid` which now copy a bunch of strings unnecessarily due to the other fix. I plan to tackle that but I wanted to get this merged first because it keeps tripping up the CI.)

Reply:
Sorry, I wasn't implying for you to go out of your way to run benchmarks.
I agree with your view that the cost in performance here is negligible.
However, I also think it's important to document why a tradeoff was being made and just "to have less unsafe code", without a word about performance, didn't feel right for me. (Also it might help people in the future looking for performance  improvements to not simply revert these changes.)

Reply:
> Sorry, I wasn't implying for you to go out of your way to run benchmarks.

Don't worry, I ran them for my own curiosity. It's a good habit to test assumptions.

> However, I also think it's important to document why a tradeoff was being made and just "to have less unsafe code", without a word about performance, didn't feel right for me. (Also it might help people in the future looking for performance improvements to not simply revert these changes.)

That's fair, I'll add a comment.